y---
title: "2- Summarize and Visualize feature count"
output: html_document
---

```{r renv_setup, include = TRUE}
#renv::init() # only run once when the project is created.
renv::restore() # run every time when a different computer is used (practically it is Ok to run every time)
#renv::snapshot() # run every time in the end if there are new packages installed (practically it is Ok to run every time in the end)
```
```{r setup, include = TRUE}

knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(ggrepel)
```


## Preparation
1.Import featureCounts outputs
```{sh}
# header, copy the terminal outputs
sed '2q;d' ../data/featureCounts/featureCounts_out | sed 's/data\/mapped_STAR\/Aligned.out.bam://g'
```
```{r}
avg_fragment_len = 294 # this is from STAR output, for PE150
# Import feature count outputs
ft_0 <- read.table("../data/featureCounts/featureCounts_out", header = T, sep = "\t", check.names = F)
colnames(ft_0) <- colnames(ft_0) %>% gsub("data\\/mapped_STAR\\/Aligned\\.out\\.bam\\:CSR2019-", "CSR2019\\.", .)

```

2. Import metadata
```{r}
mt_raw <- read.table("../data/Metadata_RNAseq_06072021-jw.txt", header = T, sep = "\t")

# metadata integrity check
sample_id = mt_raw %>% filter(Updates=="RNASeqFinished") %>% pull(TubeID)
intersect(sample_id, colnames(ft_0)) %>% length

mt_0 <- mt_raw %>% filter(Updates=="RNASeqFinished")
```

3. Gene annotation
```{sh}
# extract gene information from the gtf file
awk -F"\t" '{if($3=="gene"){print $9}}' gencode.vM27.primary_assembly.annotation.gtf > gencode.vM27.metadata.primary_assembly.annotation.col9
```
```{r}
gene_fixer <- function(str){
    str = str[str != ""]
    str_ = str %>% str_split(., " ", simplify = T)
    str_[, 2] = gsub("\"|;$", "", str_[, 2])
    str_o = t(str_[, 2]) %>% as.data.frame()
    colnames(str_o) = str_[, 1]
    str_o = as_tibble(str_o, .name_repair = function(x){make.names(x, unique = T)})
    return(str_o)
}
gene_raw = data.frame(raw = readLines("../data/reference/gencode.vM27.metadata.primary_assembly.annotation.col9"))
gene_tbl = gene_raw$raw %>% str_split(., "; +", simplify = T) %>% split(., seq(nrow(.))) %>% lapply(., "gene_fixer") %>% data.table::rbindlist(use.names = T, fill = T)
```

9. Misc
```{r}
BM_colors = c("#377EB8", "#4DAF4A", "#E41A1C", "purple")
names(BM_colors) = c("VF", "CSR", "CS", "VD")
```

## Differential analysis

### Jun-2021-proposal
Here we use DESeq2 package and only compare VF and CS
```{r}

library(DESeq2)
library(apeglm)

mt_wk = mt_0 %>% filter(BirthMode %in% c("VF", "CS"))
ft_wk = ft_0 %>% select(1:5, mt_wk$TubeID)

bodysites = c("Ileum", "Liver")

plot_volcano=function(dat, x = "log2FoldChange", y = "padj", color = "isDE", size = "baseMean"){
    g = ggplot(data = dat, aes(x = .data[[x]], y = -log10(.data[[y]]), color = .data[[color]])) + 
        geom_point(aes(size = .data[[size]])) + 
        #geom_vline(xintercept=c(-1, 1), col="red") + 
        geom_hline(yintercept=-log10(0.1), col="red") +
        labs(x = "Log2 Fold Change", y = "-log10(adjusted pValue)") + 
        theme_minimal()
    g
}

for (ss in bodysites){
    # ss = "Ileum"
    # prepare DESeq2 dataset
    ## metadata, rownames are sample name
    coldata = mt_wk %>% filter(BodySite==ss) %>% select(TubeID, BirthMode, Sex) %>% mutate(across(.col = -1, .fns = factor)) %>% column_to_rownames(var = "TubeID")
    nsamples = nrow(coldata)
    ## count matrix, rownames are gene names, colnames are sample names
    cts = ft_wk %>% select(1, rownames(coldata)) %>% column_to_rownames(var = "Geneid")
    ## DESeq2 dataset
    dds <- DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = ~ BirthMode + Sex)
    dds$BirthMode <- relevel(dds$BirthMode, ref = "VF")
    
    # prefiltering to increase speed and save memory, more strict filtering is automatically applied later
    keep <- rowSums(counts(dds)) >= 10 # genes with at least 10 reads across all samples
    dds <- dds[keep, ]
    
    # running DE analysis
    dds_run <- DESeq(dds)
    res <- results(dds_run, contrast = c("BirthMode", "CS", "VF"))
    
    ## log fold change chrinkage (explanations see https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html#:~:text=Shrunken%20log2%20foldchanges%20(LFC),High%20dispersion%20values)
    resLFC <- lfcShrink(dds_run, coef = "BirthMode_CS_vs_VF", type = "apeglm", res = res)
    res_dat <- res %>% data.frame %>% mutate(padj = ifelse(is.na(padj), 1, padj), isDE = ifelse(padj<=0.1, "Yes", "No")) # explanations of NA in padj can be found https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA and https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#i-want-to-benchmark-deseq2-comparing-to-other-de-tools.
    
    p_vol <- ggplot(data = res_dat %>% mutate(baseMean2 = cut(baseMean, breaks = c(0, 1000, 10000, Inf))), aes(x = log2FoldChange, y = -log10(padj), color = isDE)) + 
        geom_point(aes(size = baseMean2)) + 
        #geom_vline(xintercept=c(-1, 1), col="red") + 
        geom_hline(yintercept=-log10(0.1), col="red") +
        scale_color_manual(values = c("No" = "gray70", "Yes" = "red")) +
        scale_size_manual(values = c(1, 2, 4), guide = "bins") +
        theme_minimal(base_size = 12) +
        labs(x = "Log2 Fold Change", y = "-log10(adjusted p Value)", color = "Differential\nExpressed", size = "Normalized\nMean Count", title = paste0("Volcano plots showing expression changes comparing CS to VF in ", ss))
    print(p_vol)
    



}
```

